Initial working implementation of fuzzy matching

davezuch opened the pull request.

davezuch committed:
Initial implementation of algorithm. Pretty much there, need to add
tests

davezuch committed:
Cleaned up warnings and unused imports

davezuch committed:
Improved heuristics for ranking matches after feedback from @sanket

davezuch committed:
Improved heuristics for ranking matches after feedback from @SanketSabnis

davezuch committed:
Merge branch 'develop' of github.com:citizennet/purescript-text-fuzzy into develop

davezuch committed:
Implemented fuzzy matching for polymorphic types

davezuch committed:
Converted some data types to newtypes, renamed match' to matchStr

davezuch committed:
Updated algorithm to recursively check first entire pattern,
then if that's not matched, individual words in the pattern, then
if any of those words aren't matched, each individual characters
in the unmatched word, adding the severest penalty each time a match
is not found

Cleaned up and replaced `Map String String` and `StrMap String`

Added bunch of tests

davezuch committed:
Removed Main.purs

davezuch committed:
Added generics rep dependency

davezuch committed:
Added CircleCI config

davezuch committed:
Added CircleCI badge

thomashoneyman commented on src/Data/Fuzzy/Fuzzy.purs:
```diff
@@ -0,0 +1,222 @@
+instance showRank :: Show Rank where show = genericShow
+instance ordRank :: Ord Rank where compare = genericCompare
+
+instance semiringRank :: Semiring Rank where
```
This isn't a law-abiding Semiring instance (see https://pursuit.purescript.org/packages/purescript-prelude/3.0.0/docs/Data.Semiring#t:Semiring).

For instance, 

**Annihilation:**  `zero * a = a * zero = zero`

yet: `mul (Rank 0 1 0 1 0 1) zero = None`

Almost all the laws are fine (associativity, right- and left- distributivity over multiplication, etc.) but the zero provides the problem. Is it possible to give this a law-abiding Semiring instance (and thus opt-in to the guarantees of the typeclass)?

davezuch replied:
If `zero == None` then doesn't that abide by the annihilation law?

MonoidMusician replied:
It does feel a little strange to me, but I can’t come up with any counterexamples ... but perhaps a `Monoid` instance would be sufficient ? `scoreDistance` could be refactored to not use multiplication (`s/1/d/`), and it looks like the rest of it uses addition anyways, which would indicate to me that the semantics of it fit well with an (additive) monoid. Just because it’s numeric doesn’t mean multiplication makes sense ;) (e.g. 2D or 3D Vectors can be combined additively, but not really multiplied ... you can scale them with a group action though!)

davezuch replied:
Hey @MonoidMusician thanks for the feedback! I used [purescript-quickcheck-laws](https://github.com/garyb/purescript-quickcheck-laws) at your suggestion and it does seem* to abide by the `Semiring` laws, but I rather like your idea to use `Monoid` instance instead, and might just go with that

\* I say it seems to abide by the laws bc I'm not sure it manages full coverage. My instance for `Arbitrary` is defined as following:

```
instance arbitraryRank :: Arbitrary Rank
  where
    arbitrary = Rank
      <$> arbitrary
      <*> arbitrary
      <*> arbitrary
      <*> arbitrary
      <*> arbitrary
      <*> arbitrary
```
But that will never generate `None` (`Rank` is defined as `newtype Rank = Rank Int Int Int Int Int Int | None`) so I'm not certain my coverage is complete (although `zero = None` and assumedly the `Semiring` tests involve `zero`)

MonoidMusician replied:
sure, I think i can see how that makes sense ... btw https://pursuit.purescript.org/packages/purescript-quickcheck/4.6.1/docs/Test.QuickCheck.Gen#v:frequency if you want ;)

thomashoneyman commented on src/Data/Fuzzy/Fuzzy.purs:
```diff
@@ -0,0 +1,222 @@
+module Data.Fuzzy where
```
This applies to the module in general -- I could really use some comments to explain what things like `Fuzzy a` and `Result` are. It's especially true for things that are string-based like `Result`. Nothing too fancy, but a little type and function documentation inline will help me when say using this in the typeahead to ensure I know how to use it properly.

thomashoneyman replied:
I'm finding it tough to understand where the 'forgiving' filtering / ranking is happening and how to modify that. Is there a specific place in the code where this is happening and say a knob can be turned to adjust it?

thomashoneyman commented on src/Data/Fuzzy/Fuzzy.purs:
```diff
@@ -0,0 +1,222 @@
+instance eqPos :: Eq Pos where eq = genericEq
+instance showPos :: Show Pos where show = genericShow
+
+data Rank = Rank Int Int Int Int Int Int | None
```
I know we've discussed this in person, but I could use a comment here briefly noting how the ranking works (0 0 0 0 0 0 is an exact match, higher numbers are worse, `None` indicates nothing matched at all). Something quick and dirty is fine.

davezuch replied:
Yeah this was a premature PR, I do intend to go through and add thorough documentation, should've done so beforehand

thomashoneyman commented on src/Data/Fuzzy/Fuzzy.purs:
```diff
@@ -0,0 +1,222 @@
+
+type Result = Array (Either String String)
+
+match :: ∀ a. Boolean -> (a -> StrMap String) -> String -> a -> Maybe (Fuzzy a)
```
It's worth explaining what the `extract` function is up to here, mostly for helping us remember what we need to supply ("This takes an arbitrary data structure and turns it into a map of string keys to string values, where the string value is what will actually be fuzzy-matched against") -- unless that's incorrect!

Something we can glance at later on and remember what to supply and how it'll be matched against.

thomashoneyman commented on test/Main.purs:
```diff
@@ -1,9 +1,233 @@
+derive instance genericTestRecord :: Generic TestRecord _
+instance eqTestRecord :: Eq TestRecord where eq = genericEq
+instance showTestRecord :: Show TestRecord where show = genericShow
+
```
This could be a good place to check instances of classes with explicit laws like `Semiring` using property-based testing like this:

https://pursuit.purescript.org/packages/purescript-strongcheck/3.1.0/docs/Test.StrongCheck

MonoidMusician replied:
or https://github.com/garyb/purescript-quickcheck-laws ;)

davezuch committed:
Updated some names: Rank/score -> Distance, Depth -> Scope
Changed Distance from instance of Semiring to Monoid
Added law checks to ensure compliance with Monoid laws
Added ratio key to Fuzzies that denotes ratio of matched chars
Removed Maybes, now match functions will always return a Fuzzy/FuzzyStr

davezuch committed:
Added CircleCI workflow

davezuch committed:
Cleanup

davezuch committed:
Renamed shadowed var from distance to spacing

thomashoneyman reviewed (changes requested):
This is a wonderful refactor. I still think it should get some inline documentation but it's much more understandable at a glance than before.

With those docs this is a thumbs up from me

davezuch committed:
Added documentation to types and functions

davezuch committed:
Additional comment
